File: be/src/main/java/sg/flow/repositories/user/UserRepositoryImpl.java
package sg.flow.repositories.user;

import java.sql.Connection;
import java.sql.Date;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Optional;

import org.springframework.stereotype.Repository;

import sg.flow.configs.DatabaseConnectionPool;
import sg.flow.entities.User;
import sg.flow.models.user.UpdateUserProfile;
import sg.flow.models.user.UserProfile;
import sg.flow.repositories.utils.UserQueryStore;

@Repository
public class UserRepositoryImpl implements UserRepository {
    DatabaseConnectionPool databaseConnectionPool;

    private final String SAVE = UserQueryStore.SAVE_USER;

    private final String SAVE_WITH_ID = UserQueryStore.SAVE_USER_WITH_ID;

    private final String FIND_BY_ID = UserQueryStore.FIND_USER_BY_ID;

    private final String DELETE_ALL = UserQueryStore.DELETE_ALL_USERS;

    private final String FIND_USER_PROFILE = UserQueryStore.FIND_USER_PROFILE;

    private final String FIND_USER_PREFERENCE_JSON = UserQueryStore.FIND_USER_PREFERENCE_JSON;

    private final String UPDATE_USER_PROFILE = UserQueryStore.UPDATE_USER_PROFILE;

    public UserRepositoryImpl(DatabaseConnectionPool databaseConnectionPool) {
        this.databaseConnectionPool = databaseConnectionPool;
    }

    @Override
    public User save(User entity) {
        Connection connection = databaseConnectionPool.getConnection();
        boolean hasId = entity.getId() != null;
        String queryString = hasId ? SAVE_WITH_ID : SAVE;
        int parameterStart = hasId ? 1 : 0;
        try (PreparedStatement pstm = connection.prepareStatement(queryString)) {
            if (hasId) {
                pstm.setLong(parameterStart, entity.getId());
            }
            pstm.setString(parameterStart + 1, entity.getName());
            pstm.setString(parameterStart + 2, entity.getEmail());
            pstm.setString(parameterStart + 3, entity.getIdentificationNumber());
            pstm.setString(parameterStart + 4, entity.getPhoneNumber());
            pstm.setDate(parameterStart + 5, Date.valueOf(entity.getDateOfBirth()));
            pstm.setString(parameterStart + 6, entity.getAddress());
            pstm.setString(parameterStart + 7, entity.getSettingJson());

            pstm.executeUpdate();
            return entity;
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        } finally {
            try {
                connection.close();
            } catch (SQLException e1) {
                e1.printStackTrace();
            }
        }
    }

    @Override
    public Optional<User> findById(Long id) {
        if (id == null) {
            return Optional.empty();
        }
        Connection connection = databaseConnectionPool.getConnection();
        try (PreparedStatement pstm = connection.prepareStatement(FIND_BY_ID)) {
            pstm.setLong(1, id);

            ResultSet resultSet = pstm.executeQuery();
            if (resultSet.next()) {
                User user = User.builder()
                        .id(resultSet.getInt("id"))
                        .name(resultSet.getString("name"))
                        .email(resultSet.getString("email"))
                        .identificationNumber(resultSet.getString("identification_number"))
                        .phoneNumber(resultSet.getString("phone_number"))
                        .dateOfBirth(resultSet.getDate("date_of_birth").toLocalDate())
                        .address(resultSet.getString("address"))
                        .settingJson(resultSet.getString("setting_json"))
                        .build();
                return Optional.of(user);
            }
        } catch (SQLException e) {
            e.printStackTrace();
            return Optional.empty();
        } finally {
            try {
                connection.close();
            } catch (SQLException e1) {
                e1.printStackTrace();
            }
        }
        return null;
    }

    @Override
    public boolean deleteAll() {
        Connection connection = databaseConnectionPool.getConnection();
        try (PreparedStatement pstm = connection.prepareStatement(DELETE_ALL)) {
            pstm.executeUpdate();
            return true;
        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        } finally {
            try {
                connection.close();
            } catch (SQLException e1) {
                e1.printStackTrace();
            }
        }
    }

    @Override
    public UserProfile getUserProfile(Integer id) {
        Connection connection = databaseConnectionPool.getConnection();
        try (PreparedStatement pstm = connection.prepareStatement(FIND_USER_PROFILE)) {
            pstm.setInt(1, id);

            ResultSet resultSet = pstm.executeQuery();
            if (resultSet.next()) {
                UserProfile userProfile = new UserProfile(
                        resultSet.getString("name"),
                        resultSet.getString("email"),
                        resultSet.getString("phone_number"),
                        resultSet.getDate("date_of_birth").toLocalDate(),
                        resultSet.getString("identification_number"),
                        resultSet.getString("address"));
                return userProfile;
            }
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        } finally {
            try {
                connection.close();
            } catch (SQLException e1) {
                e1.printStackTrace();
            }
        }
        return null;
    }

    @Override
    public String getUserPreferenceJson(Integer userId) {
        Connection connection = databaseConnectionPool.getConnection();
        try (PreparedStatement pstm = connection.prepareStatement(FIND_USER_PREFERENCE_JSON)) {
            pstm.setInt(1, userId);

            ResultSet resultSet = pstm.executeQuery();
            if (resultSet.next()) {
                return resultSet.getString("setting_json");
            }
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        } finally {
            try {
                connection.close();
            } catch (SQLException e1) {
                e1.printStackTrace();
            }
        }
        return null;
    }

    @Override
    public UserProfile updateUserProfile(Integer userId, UpdateUserProfile updateUserProfile) {
        Connection connection = databaseConnectionPool.getConnection();
        UserProfile userProfile = this.getUserProfile(userId);
        this.mergeUserProfile(userProfile, updateUserProfile);
        if (userProfile == null) {
            return null;
        }
        try (PreparedStatement pstm = connection.prepareStatement(UPDATE_USER_PROFILE)) {
            pstm.setString(1, userProfile.getEmail());
            pstm.setString(2, userProfile.getPhoneNumber());
            pstm.setString(3, userProfile.getAddress());

            pstm.setInt(4, userId);

            pstm.executeUpdate();

            return userProfile;
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        } finally {
            try {
                connection.close();
            } catch (SQLException e1) {
                e1.printStackTrace();
            }
        }
    }

    private void mergeUserProfile(UserProfile originalUserProfile, UpdateUserProfile userProfile) {
        if (userProfile.getEmail() != null) {
            originalUserProfile.setEmail(userProfile.getEmail());
        }
        if (userProfile.getPhoneNumber() != null) {
            originalUserProfile.setPhoneNumber(userProfile.getPhoneNumber());
        }
        if (userProfile.getAddress() != null) {
            originalUserProfile.setAddress(userProfile.getAddress());
        }
    }
}

File: be/src/main/kotlin/sg/flow/repositories/user/UserRepositoryImpl.kt
package sg.flow.repositories.user

import io.r2dbc.postgresql.codec.Json
import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.reactive.asFlow
import java.time.LocalDate
import kotlinx.coroutines.reactive.awaitFirstOrNull
import kotlinx.coroutines.reactive.awaitSingle
import org.slf4j.LoggerFactory
import org.springframework.r2dbc.core.DatabaseClient
import org.springframework.r2dbc.core.awaitRowsUpdated
import org.springframework.stereotype.Repository
import sg.flow.entities.User
import sg.flow.grpc.UserIdAndPasswordHash
import sg.flow.models.user.UpdateUserProfile
import sg.flow.models.user.UserProfile
import sg.flow.repositories.utils.UserQueryStore

@Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
@Repository
class UserRepositoryImpl(private val databaseClient: DatabaseClient) : UserRepository {

    val logger = LoggerFactory.getLogger(UserRepositoryImpl::class.java)

    override suspend fun save(entity: User): User {
        val hasId = entity.id != null
        val queryString = if (hasId) UserQueryStore.SAVE_USER_WITH_ID else UserQueryStore.SAVE_USER

        if (hasId) {
            var spec = databaseClient
                    .sql(queryString)
                    .bind(0, entity.id)
                    .bind(1, entity.name)
                    .bind(2, entity.email)
                    .bind(3, entity.identificationNumber)
                    .bind(4, entity.phoneNumber)

            spec = if (entity.dateOfBirth == null) {
                spec.bindNull(5, LocalDate::class.java)
            } else {
                spec.bind(5, entity.dateOfBirth)
            }

            try {
                spec.bind(6, entity.address)
                    .bind(7, entity.passwordHash)
                    .bind(8, Json.of(entity.settingJson))
                    .fetch()
                    .awaitRowsUpdated()
            } catch (e: Exception) {
                e.printStackTrace()
                logger.error("Error while saving user: ${entity.id}")
            }

        } else {
            var spec = databaseClient
                    .sql(queryString)
                    .bind(0, entity.name)
                    .bind(1, entity.email)
                    .bind(2, entity.identificationNumber)
                    .bind(3, entity.phoneNumber)

            spec = if (entity.dateOfBirth == null) {
                spec.bindNull(4, LocalDate::class.java)
            } else {
                spec.bind(4, entity.dateOfBirth)
            }

            try {
                spec.bind(5, entity.address)
                    .bind(6, entity.passwordHash)
                    .bind(7, Json.of(entity.settingJson))
                    .fetch()
                    .awaitRowsUpdated()
            } catch (e: Exception) {
                e.printStackTrace()
                logger.error("Error while saving user with name: ${entity.name}")
            }

        }

        return entity
    }

    override suspend fun findById(id: Long): User? {
        return databaseClient
                .sql(UserQueryStore.FIND_USER_BY_ID)
                .bind(0, id)
                .map { row ->
                    User(
                            id = row.get("id", Int::class.java)!!,
                            name = row.get("name", String::class.java)!!,
                            email = row.get("email", String::class.java)!!,
                            identificationNumber =
                                    row.get("identification_number", String::class.java)!!,
                            phoneNumber = row.get("phone_number", String::class.java)!!,
                            dateOfBirth = row.get("date_of_birth", LocalDate::class.java)!!,
                            address = row.get("address", String::class.java)!!,
                            settingJson = row.get("setting_json", String::class.java)!!,
                            gender_is_male = getGenderFromDBText(row.get("gender", String::class.java)!!)
                    )
                }
                .one()
                .awaitFirstOrNull()
    }



    override suspend fun deleteAll(): Boolean {
        return try {
            databaseClient.sql(UserQueryStore.DELETE_ALL_USERS).fetch().awaitRowsUpdated()
            true
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }

    override fun getAllUserIds(): List<Int> {
        return runCatching {
            val sql = UserQueryStore.FIND_ALL_USER_IDS

            databaseClient.sql(sql)
                .map { row -> row.get("id", Int::class.java) }
                .all()
                .collectList()              // Mono<List<Int?>>
                .map { it.filterNotNull() } // Mono<List<Int>>
                .block() ?: emptyList()

        }.onFailure { e ->
            e.printStackTrace()
            logger.error("Failed to fetch all user IDs")
        } .getOrElse { emptyList() }
    }

    override suspend fun getUserProfile(id: Int): UserProfile {
        try {
            val profile = databaseClient
                .sql(UserQueryStore.FIND_USER_PROFILE)
                .bind(0, id)
                .map { row ->
                    UserProfile(
                        id = id,
                        name = row.get("name", String::class.java)!!,
                        email = row.get("email", String::class.java)!!,
                        phoneNumber = row.get("phone_number", String::class.java)!!,
                        dateOfBirth = row.get("date_of_birth", LocalDate::class.java),
                        identificationNumber =
                            row.get("identification_number", String::class.java)!!,
                        settingJson = row.get("setting_json", String::class.java)!!,
                        genderIsMale = getGenderFromDBText(row.get("gender", String::class.java))
                    )
                }
                .one()
                .awaitFirstOrNull()

            if (profile == null) {
                logger.error("Couldn't find profile with id: $id")
                throw RuntimeException()
            }
            return profile
        } catch (e: Exception) {
            e.printStackTrace()
            logger.error("Failed to get user profile of id: $id")
            throw RuntimeException("User with id $id not found")
        }
    }

    override suspend fun getUserPreferenceJson(userId: Int): String {
        return databaseClient
                .sql(UserQueryStore.FIND_USER_PREFERENCE_JSON)
                .bind(0, userId)
                .map { row -> row.get("setting_json", String::class.java) ?: "{}" }
                .one()
                .awaitFirstOrNull()
                ?: throw RuntimeException("User with id $userId not found")
    }

    override suspend fun updateUserProfile(
        userId: Int,
        userProfile: UpdateUserProfile
    ): UserProfile {

        // 2) bind & execute, without .fetch()
        var rowsUpdated: Long = -1L;
        try {
            rowsUpdated = databaseClient
                .sql(UserQueryStore.UPDATE_USER_PROFILE)
                .bind(0, userProfile.email ?: "")
                .bind(1, userProfile.phoneNumber ?: "")
                .bind(2, Json.of(userProfile.settingsJson))
                .bind(3, userId)
                .fetch()
                .rowsUpdated()
                .awaitSingle()
        } catch (e: Exception) {
            e.printStackTrace()
            logger.error("Error while updating user profile for user ID $userId")
        }


        if (rowsUpdated == 0L) {
            throw RuntimeException("User with id $userId not found")
        }
        return getUserProfile(userId)
    }

    override suspend fun checkUserExists(email: String): Boolean {
        val sql = UserQueryStore.CHECK_USER_EXISTS_BY_EMAIL

        val result = databaseClient.sql(sql)
            .bind(0, email)
            .map { row ->
                row.get("count", Int::class.java) == 1
            }
            .one()
            .awaitFirstOrNull()

        return result ?: false
    }

    override suspend fun getUserIdByEmail(email: String): Int {
        val sql = UserQueryStore.FIND_USER_ID_BY_EMAIL

        val result = databaseClient.sql(sql)
            .bind(0, email)
            .map {row ->
                row.get("id", Int::class.java)
            }
            .one()
            .awaitFirstOrNull()

        return result ?: -1
    }

    override suspend fun getUserIdAndPasswordHashWithEmail(email: String): UserIdAndPasswordHash {
        val sql = UserQueryStore.FIND_USER_ID_AND_PASSWORD_HASH_BY_EMAIL

        val userIdAndPasswordHash = databaseClient.sql(sql)
            .bind(0, email)
            .map { row ->
                UserIdAndPasswordHash(
                    row.get("id", Int::class.java) ?: -1,
                    row.get("password_hash", String::class.java) ?: ""
                )

            }
            .one()
            .awaitFirstOrNull()

        return userIdAndPasswordHash ?: UserIdAndPasswordHash(-1, "")
    }

    override suspend fun markUserEmailVerified(email: String): Boolean {
        val sql = UserQueryStore.MARK_USER_EMAIL_VERIFIED

        val result = databaseClient.sql(sql)
            .bind(0, email)
            .bind(1, true)
            .fetch()
            .awaitRowsUpdated()

        return result == 1L
    }

    override suspend fun fetchIsUserEmailVerified(email: String): Boolean {
        val sql = UserQueryStore.FIND_USER_EMAIL_VERIFIED

        val result = databaseClient.sql(sql)
            .bind(0, email)
            .map {row ->
                row.get("is_email_verified", Boolean::class.java)
            }
            .one()
            .awaitFirstOrNull()

        if (result != null) {
            return result
        }
        return false
    }

    override suspend fun canLinkBank(userId: Int): Boolean {
        val sql = UserQueryStore.CHECK_USER_CAN_LINK_BANK;

        val result = databaseClient.sql(sql)
            .bind(0, userId)
            .map { row ->
                val dateOfBirth = row.get("date_of_birth", LocalDate::class.java)
                val gender = getGenderFromDBText(row.get("gender", String::class.java))

                if (dateOfBirth == null || gender == null) {
                    false
                } else {
                    true
                }
            }
            .one()
            .awaitFirstOrNull()

        if (result != null) {
            return result
        }
        return false
    }

    override suspend fun setConstantUserFields(
        userId: Int,
        dateOfBirth: LocalDate,
        genderIsMale: Boolean
    ): Boolean {
        val sql = UserQueryStore.SET_CONSTANT_USER_FIELDS

        val result = databaseClient.sql(sql)
            .bind(0, userId)
            .bind(1, dateOfBirth)
            .bind(2, getDBGenderTextFromGenderBoolean(genderIsMale))
            .fetch()
            .awaitRowsUpdated()

        return result == 1L
    }

    companion object {
        fun getGenderFromDBText(genderText: String): Boolean? {
            if (genderText.isEmpty()) {
                return null
            }

            if (genderText == "MALE") {
                return true;
            } else if (genderText == "FEMALE") {
                return false
            }
            return null
        }

        fun getDBGenderTextFromGenderBoolean(genderBool: Boolean) : String {
            if (genderBool) {
                return "MALE"
            }
            return "FEMALE"
        }
    }
}
